modules/fedimint-mint-client/src/redemption.rs

```diff
@@ -0,0 +1,279 @@
+use fedimint_client::sm::{ClientSMDatabaseTransaction, OperationId, State, StateTransition};
+use fedimint_client::transaction::{ClientInput, TransactionBuilder};
+use fedimint_client::DynGlobalClientContext;
+use fedimint_core::core::DynInput;
+use fedimint_core::encoding::{Decodable, Encodable};
+use fedimint_core::{TieredMulti, TransactionId};
+use fedimint_mint_common::MintInput;
+
+use crate::{MintClientContext, SpendableNote};
+
+// TODO: add retry with valid subset of e-cash notes
+/// State machine managing the e-cash redemption process related to a mint
+/// input.
+///
+/// ```mermaid
+/// graph LR
+///     classDef virtual fill:#fff,stroke-dasharray: 5 5
+///
+///     Created -- containing tx accepted --> Success
+///     Created -- containing tx rejected --> Refund
+///     Refund -- refund tx rejected --> Error
+///     Refund -- refund tx accepted --> RS[Refund Success]
+/// ```
+#[derive(Debug, Clone, Eq, PartialEq, Decodable, Encodable)]
+pub enum MintRedemptionStates {
```

Comment:
Should this be named "redemption" or just "input". Redemption here is just a special case of tracking input state. What we're doing is tracking "mint input" state, not redemption.

This makes me thing that for issuance it's less obviously, but also applies. We're tracking "mint output" state, not "issuance". "Issuance" is just a higher level interpretation of what is happening.